### ISSUE TITLE
Avg loss straight to zero

### ISSUE THREAD

**Issue by JoeCool90** [sowson/darknet#16]
Hi, thanks for doing this opencl fork.  I've got further on getting things working for my RX 580 than others.

I needed to do sudo ldconfig to get it to run, and then needed to put:

*output = clLinkProgram(opencl_contexts[opencl_device_id_t], opencl_device_ct_t, opencl_devices, "", 1, prog, NULL, NULL, &clErr);

in opencl.c

I'm now to train on the voc data set, following AlexeyAB's tutorial.  The problem is, the average loss function is going straight to zero.

Here is the output from the first iterations, thanks:

Loading weights from darknet53.conv.74...Done!
Learning Rate: 0.001, Momentum: 0.9, Decay: 0.0005
Resizing
448
Loaded: 0.000054 seconds
Region 82 Avg IOU: 0.145264, Class: 0.557355, Obj: 0.292876, No Obj: 0.494451, .5R: 0.000000, .75R: 0.000000,  count: 9
Region 94 Avg IOU: 0.149600, Class: 0.676368, Obj: 0.355952, No Obj: 0.513023, .5R: 0.142857, .75R: 0.000000,  count: 14
Region 106 Avg IOU: 0.007626, Class: 0.401388, Obj: 0.633496, No Obj: 0.471143, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 82 Avg IOU: 0.257950, Class: 0.426642, Obj: 0.322373, No Obj: 0.493180, .5R: 0.200000, .75R: 0.000000,  count: 10
Region 94 Avg IOU: 0.137465, Class: 0.400465, Obj: 0.394822, No Obj: 0.512317, .5R: 0.000000, .75R: 0.000000,  count: 7
Region 106 Avg IOU: 0.202138, Class: 0.586369, Obj: 0.364813, No Obj: 0.473299, .5R: 0.250000, .75R: 0.000000,  count: 4
Region 82 Avg IOU: 0.291640, Class: 0.488285, Obj: 0.518656, No Obj: 0.493739, .5R: 0.166667, .75R: 0.000000,  count: 6
Region 94 Avg IOU: 0.093404, Class: 0.393986, Obj: 0.703347, No Obj: 0.514841, .5R: 0.000000, .75R: 0.000000,  count: 9
Region 106 Avg IOU: 0.379017, Class: 0.924878, Obj: 0.687865, No Obj: 0.475272, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 82 Avg IOU: 0.285344, Class: 0.346855, Obj: 0.317452, No Obj: 0.492233, .5R: 0.142857, .75R: 0.000000,  count: 7
Region 94 Avg IOU: 0.130376, Class: 0.506110, Obj: 0.634440, No Obj: 0.512611, .5R: 0.000000, .75R: 0.000000,  count: 5
Region 106 Avg IOU: 0.119781, Class: 0.555439, Obj: 0.783421, No Obj: 0.472111, .5R: 0.000000, .75R: 0.000000,  count: 2
Region 82 Avg IOU: 0.286759, Class: 0.532216, Obj: 0.521017, No Obj: 0.493246, .5R: 0.100000, .75R: 0.000000,  count: 10
Region 94 Avg IOU: 0.137215, Class: 0.514404, Obj: 0.459499, No Obj: 0.513389, .5R: 0.000000, .75R: 0.000000,  count: 5
Region 106 Avg IOU: 0.153096, Class: 0.403162, Obj: 0.553437, No Obj: 0.470921, .5R: 0.000000, .75R: 0.000000,  count: 4
Region 82 Avg IOU: 0.223061, Class: 0.484147, Obj: 0.489988, No Obj: 0.492903, .5R: 0.117647, .75R: 0.000000,  count: 17
Region 94 Avg IOU: 0.040685, Class: 0.678096, Obj: 0.464522, No Obj: 0.513007, .5R: 0.000000, .75R: 0.000000,  count: 5
Region 106 Avg IOU: 0.351594, Class: 0.760295, Obj: 0.369242, No Obj: 0.469966, .5R: 0.000000, .75R: 0.000000,  count: 2
Region 82 Avg IOU: 0.209237, Class: 0.635235, Obj: 0.570351, No Obj: 0.493333, .5R: 0.000000, .75R: 0.000000,  count: 10
Region 94 Avg IOU: 0.220527, Class: 0.463515, Obj: 0.465429, No Obj: 0.514360, .5R: 0.000000, .75R: 0.000000,  count: 5
Region 106 Avg IOU: 0.260902, Class: 0.483188, Obj: 0.516514, No Obj: 0.472673, .5R: 0.000000, .75R: 0.000000,  count: 4
Region 82 Avg IOU: 0.248426, Class: 0.553127, Obj: 0.496878, No Obj: 0.492399, .5R: 0.000000, .75R: 0.000000,  count: 11
Region 94 Avg IOU: 0.355078, Class: 0.793197, Obj: 0.322255, No Obj: 0.512674, .5R: 0.500000, .75R: 0.000000,  count: 2
Region 106 Avg IOU: 0.063738, Class: 0.852796, Obj: 0.704434, No Obj: 0.472485, .5R: 0.000000, .75R: 0.000000,  count: 1
1: 1132.090576, 1132.090576 avg, 0.000000 rate, 109.115808 seconds, 64 images

**Comment by JoeCool90** [comment]
Oops, I was looking at the rate column.  Don't know if 0.0000 is a problem but obviously not what I was thinking.

**Comment by sowson** [comment]
Hi @JoeCool90 thanks, I am still looking for a solution why loss error not going as far low as on the CUDA version... it is harder to find then I expected. What is the result after 100 or 1000 steps? Is it going down? That is Yolo3 Training am I right? Thanks!

**Comment by JoeCool90** [comment]
I only tested to the yolo3 on voc dataset for a bit, but now I have switched to a custom training set (yolov3).  So far, after 857 iterations it has come steadily down to under 1.0.  This has taken nearly 24hrs on a gtx580 with one class.  Not sure how that compares, but so far it seems to be working.

**Comment by yaoanderson** [comment]
@JoeCool90, could you please upload your yolov3.cfg based on your training, and i can not train my network under 1.0 loss, thanks.

**Comment by JoeCool90** [comment]
Attached (was cfg, but needed to change to txt to upload).

It follows the instructions on [https://github.com/AlexeyAB/darknet](url) including the bit about anchors.  If you're getting near to 1.0 then maybe that's as good as you're going to get.  Apparently up to 3.0 may be obtained if the model is large/difficult.

[yolo-obj.txt](https://github.com/sowson/darknet/files/3240716/yolo-obj.txt)

**Comment by yaoanderson** [comment]
@JoeCool90 Thanks.

**Comment by JoeCool90** [comment]
So far this is just a test.  The learning rate was in the default yolov3 AlexeyAB file (and I presume the others?), I haven't touched it.

The link above will tell you how to do the anchors, you basically run a programme and it coughs them out.  See towards the bottom in the tips section.

The random=0 thing was probably when I was experiencing crashes with other forks and trying a few suggestions.  Next time I'll put it back to 1

**Comment by yaoanderson** [comment]
@JoeCool90 But AlexeyAB/darknet is not based on opencl ?! We discussed all based on sowson's darknet, right ?

**Comment by JoeCool90** [comment]
AlexeyAB has the best (kinda only) tutorial.  But all the forks, from the original pjreddie, to AlexeyAB, to Sowson, etc... should produce the same results more or less.  The operations and computation should be the same, just different hardware compatibility

**Comment by yaoanderson** [comment]
@joecool90 you are right. sowson have supported opengl, so I choose it, but so far I can not get result what I want, and I am waiting for the research and feedback from sowson.

**Comment by JoeCool90** [comment]
I've just finished my first training (2000 iterations).  Unfortunately the results are nonsense.  I'm going to have a look to see if I prepared my data wrong.  Otherwise, I'll try with the VOC dataset, and if that all works, not sure what the problem is.

**Comment by JoeCool90** [comment]
Just checked the data via another annotation tool and that was fine. Also tried the pre-trained coco weights and that worked.  So something messed up in my training.

**Comment by sowson** [comment]
Hi, please check the fix I just committed for Loss Error Fix. Thanks! :D

**Comment by yaoanderson** [comment]
@joeCool90 Is it working fine now ?  @sowson How about my question?

**Comment by sowson** [comment]
@yaoanderson can you please remind me your question? Thanks!

**Comment by JoeCool90** [comment]
@yaoanderson  I'm running another training session.  I tried with random = 1 but the computer more or less seized up and then the programme crashed with several opencl memory errors.

**Comment by sowson** [comment]
@JoeCool90 what errors you got? You may try to reduce the factor that is batch/subdivision values from batch=64 subdivisions=8 to for example batch=64 subdivisions=16 or batch=32 subdivisions=16 Thanks!

**Comment by JoeCool90** [comment]
Resizing
576
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE
could create buffer on device. error: CL_MEM_OBJECT_ALLOCATION_FAILURE

I'll try, thanks

**Comment by sowson** [comment]
I am testing now batch=16 subdivisions=8 to check all changes I did and committed here. I have only batch / subdivisions = 8 and 8 is only possible and works also batch=32 subdivisions=16 when such factor is still 8 on my eGPU that is RX570 4GB. Hope that helps. How much memory @JoeCool90  you have?

**Comment by yaoanderson** [comment]
@sowson About my open issue convergent slow or not issue

**Comment by JoeCool90** [comment]
@sowson I've got an RX580 8GB.  I'm trying batch=32, subdivisions=16
@yaoanderson Convergence seemed to be fine, although the final network didn't work.  Not sure what is considered fast.  It took 2.5 days to do 2000 iterations, although oddly it only saved up to 1000 and then the final weights

**Comment by sowson** [comment]
Last time I remebered that AlexeyAB teaches that the last yolo kind layer should be commented out in the configuration file. There are 3 yolo layers in Yolo3 model. So maybe you should look on that. 8GB is good and enough I think. This is only a tip I am not 100% sure about it. I will try that as well soon, commented out in CFG the 3rd yolo entry. Thanks, maybe that will help you.

**Comment by JoeCool90** [comment]
I've just had a big re-read of the tips from AlexeyAB.  I can't find anything about commenting out the third layer, it keep mentioning it to be included.  However, there are quite a few subtle detailed points.  For example, I trying to detect a small object so I should have followed the tip about that.  Also, I didn't do the masks bit with the anchor points so I'll look at that again, along with everything else.

On a positive note, change the batch/sub numbers has calmed the system down with no more sign of memory errors.

**Comment by sowson** [comment]
On my DreamPC  
[yolov3-voc.cfg.txt](https://github.com/sowson/darknet/files/3245124/yolov3-voc.cfg.txt)
Step 300
![Zrzut ekranu z 2019-06-02 21-17-24](https://user-images.githubusercontent.com/33941184/58766254-92fcf100-857c-11e9-9c95-d3542e295031.png)
:-)... Tomorrow morning I will check how it goes after more than 1k iterations.
On Yolo2 it works like this...
![Zrzut ekranu z 2019-05-24 23-53-36](https://user-images.githubusercontent.com/33941184/58766270-c8094380-857c-11e9-9615-dc592003f517.png)
On the morning I will show you how was the same picture on Yolo3...
Thanks!

**Comment by sowson** [comment]
On 1100 steps I have.
![Zrzut ekranu z 2019-06-03 04-22-23](https://user-images.githubusercontent.com/33941184/58772341-bba2dc00-85b8-11e9-8498-3fe8cbcad576.png)
And I tested how it works?:
![Zrzut ekranu z 2019-06-03 04-28-17](https://user-images.githubusercontent.com/33941184/58772354-cfe6d900-85b8-11e9-874f-24bcbb3dfd39.png)
Looks like a little more about 2000 iterations are needed.
Thanks!
P.S. That works, so can we close this issue? It is up to you guys... :D

**Comment by sowson** [comment]
After 1500 iterations:
![Zrzut ekranu z 2019-06-03 08-10-09](https://user-images.githubusercontent.com/33941184/58779927-9d000d80-85d7-11e9-8df9-238271c44b49.png)
I think it needs some more training... at least 2000 iterations.
Thanks!

**Comment by yaoanderson** [comment]
@sowson  I am curious How about your training status about 400 rounds and 800 rounds as below data in yolov2 ?
![image](https://user-images.githubusercontent.com/4134275/58780130-976feb00-860a-11e9-954b-f64de07f3d66.png)

**Comment by sowson** [comment]
I have very busy season right now at work, I try to help you, but probably on the weekend. Earlier there is no chance to play with the C and darknet :D. Maybe someone else can help you here before that?

**Comment by JoeCool90** [comment]
I'm currently trying another run so won't be able to add anything until that finishes.

**Comment by JoeCool90** [comment]
I used smaller images for my training set, so that my targets were proportionally larger and now everything works!  Having some issues with the example python bindings but apart from that I'm happy.

**Comment by sowson** [comment]
@JoeCool90 can you help with issue 17... shame to confess but I do not know the Python at all :/.

**Comment by sowson** [comment]
Can you close this one issue if all works as expected? Thanks!

**Comment by JoeCool90** [comment]
> @JoeCool90 can you help with issue 17... shame to confess but I do not know the Python at all :/.

Sorry, I'm only a beginner on Python and I don't go near Apple.  I can't get the python example working, I think there is a memory pointer issue.  I'm going to go back to c for a bit to see if that works.  Everything from the command line is now working
